front/src/modules/timesStops/hooks/useComputedMargins.tsx

```diff
@@ -0,0 +1,73 @@
+import type { SimulationResponseSuccess } from 'applications/operationalStudies/types';
+import type { TrainScheduleResult } from 'common/api/generatedEditoastApi';
+import type { OperationalPointWithTimeAndSpeed } from 'modules/trainschedule/components/DriverTrainScheduleV2/types';
+import { interpolateValue } from 'modules/trainschedule/components/DriverTrainScheduleV2/utils';
+
+import { marginRegExValidation } from '../consts';
+import { formatDigitsAndUnit } from '../utils';
+
+let marginUnit = '%';
+
+const useComputedMargins = (
+  trainSimulation: SimulationResponseSuccess,
+  opPoint: OperationalPointWithTimeAndSpeed,
+  nextOpPoint: OperationalPointWithTimeAndSpeed,
+  selectedTrainSchedule: TrainScheduleResult,
+  pathId: string
+) => {
+  const theoreticalMarginIndex = selectedTrainSchedule?.margins?.boundaries?.findIndex(
+    (id) => id === pathId
+  );
+  if (!theoreticalMarginIndex) {
+    return ['', '', ''];
+  }
+  const theoreticalMargin = selectedTrainSchedule.margins!.values[theoreticalMarginIndex + 1];
+  const theoMarginRegexMatches = theoreticalMargin?.match(marginRegExValidation);
+  marginUnit = theoMarginRegexMatches?.[3] || marginUnit;
+  const opPointBaseTime = interpolateValue(trainSimulation.base, opPoint?.position * 1000, 'times');
+  const nextOpPointBaseTime = interpolateValue(
+    trainSimulation.base,
+    nextOpPoint?.position * 1000,
+    'times'
+  );
+  const baseDuration = nextOpPointBaseTime - opPointBaseTime;
+  const opPointprovTime = interpolateValue(
+    trainSimulation.provisional,
+    opPoint?.position * 1000,
+    'times'
+  );
+  const nextOpPointProvTime = interpolateValue(
+    trainSimulation.provisional,
+    nextOpPoint?.position * 1000,
+    'times'
+  );
+  const provisionalDuration = nextOpPointProvTime - opPointprovTime;
+  // Time between current scheduled point and next scheduled point
+  const realLostTime = provisionalDuration - baseDuration;
+  const distanceToNextOpPoint = nextOpPoint?.position - opPoint?.position;
+  let realMargin = 0;
+  let diffMargins = 0;
+  if (marginUnit === '%') {
+    realMargin = 100 * (realLostTime / baseDuration) || 0;
+    if (theoreticalMargin) {
+      const theoreticalMarginValue = Number(theoMarginRegexMatches?.[1]);
+      const theoriticalLostTime = baseDuration * (theoreticalMarginValue * 0.01);
+      diffMargins = (realLostTime - theoriticalLostTime) * 0.001;
+    }
+  } else {
+    // unit is min/100km
+    realMargin = (100 * (realLostTime / 60)) / distanceToNextOpPoint || 0;
+    if (theoreticalMargin) {
+      const theoreticalMarginValue = Number(theoMarginRegexMatches?.[1]);
+      const theoriticalLostTime = (60 * theoreticalMarginValue * distanceToNextOpPoint) / 100;
+      diffMargins = (realLostTime - theoriticalLostTime) * 0.001;
+    }
+  }
+
+  const theoreticalMarginWithUnit = formatDigitsAndUnit(theoreticalMargin);
```

Comment:
Can you re-use theoreticalMargin directly ?

Reply:
We need a non-breaking space to abide by typography rules

100 km
100 %

Reply:
Understood